src/Rel.mo

```diff
@@ -0,0 +1,178 @@
+    // and can be recomputed in linear time from the forw field.
+    //
+    // back : Trie.Trie2D<Y, X, ()> ;
+  };
+
+  public func share<X, Y>( rel : Rel<X, Y> ) : RelShared<X, Y> {
+    { forw = rel.forw ;
+      // back = rel.back ;
+    }
+  };
+
+  public func fromShare<X, Y>( rel : RelShared<X, Y>,
+                               hash_ : HashPair<X, Y>,
+                               equal_ : EqualPair<X, Y> ) : Rel<X, Y>
+  {
+    { forw = rel.forw ;
+      back = invert(rel.forw);
+      hash = hash_ ;
+      equal = equal_
+    }
+  };
+
+  public func keyOf0<X, Y>( rel : Rel<X, Y>,  x : X) : Trie.Key<X> {
+    { key = x ; hash = rel.hash.0(x) }
+  };
+
+  public func keyOf1<X, Y>( rel : Rel<X, Y>,  y : Y) : Trie.Key<Y> {
+    { key = y ; hash = rel.hash.1(y) }
+  };
+
+  public func keyOf<X, Y>( rel : Rel<X, Y>, p : (X, Y))
+    : (Trie.Key<X>, Trie.Key<Y>)
+  {
+    (keyOf0(rel, p.0),
+     keyOf1(rel, p.1))
+  };
+
+  public func empty<X, Y>( hash_ : HashPair<X, Y>,
+                           equal_ : EqualPair<X, Y>) : Rel<X, Y> {
+    {
+      forw = Trie.empty();
+      back = Trie.empty();
+      hash = hash_ ;
+      equal = equal_
+    }
+  };
+
+  public func getRelated0<X, Y>(rel : Rel<X, Y>, x : X) : Iter.Iter<Y> {
+    let t = Trie.find<X, Trie.Trie<Y, ()>>(rel.forw, keyOf0(rel, x), rel.equal.0);
+    switch t {
+      // to do -- define as Iter.empty()
+      case null { object { public func next() : ?Y { null } } };
+      case (?t) { iterAll(t) };
+    }
+  };
+
+  public func getRelated1<X, Y>(rel : Rel<X, Y>, y : Y) : Iter.Iter<X> {
+    let t = Trie.find(rel.back, keyOf1(rel, y), rel.equal.1);
+    switch t {
+      case null { object { public func next() : ?X { null } } };
+      case (?t) { iterAll(t) };
+    }
+  };
+
+  public func put<X, Y>( rel : Rel<X, Y>, p : (X, Y)) : Rel<X, Y> {
+    let k = keyOf(rel, p);
+    {
+      forw = Trie.put2D(rel.forw, k.0, rel.equal.0, k.1, rel.equal.1, ()) ;
+      back = Trie.put2D(rel.back, k.1, rel.equal.1, k.0, rel.equal.0, ()) ;
+      hash = rel.hash ;
+      equal = rel.equal ;
+    }
+  };
+
+  public func delete<X, Y>( rel : Rel<X, Y>, p : (X, Y)) : Rel<X, Y> {
+    let k = (keyOf0(rel, p.0), keyOf1(rel, p.1));
+    {
+      forw = Trie.remove2D(rel.forw, k.0, rel.equal.0, k.1, rel.equal.1).0 ;
+      back = Trie.remove2D(rel.back, k.1, rel.equal.1, k.0, rel.equal.0).0 ;
+      hash = rel.hash ;
+      equal = rel.equal ;
+    }
+  };
+
+  func invert<X, Y>(rel : Trie.Trie2D<X, Y, ()>) : Trie.Trie2D<Y, X, ()> {
+    Prelude.nyi() // to do -- for testing / upgrades sub-story
+  };
+
+  // helper for getRelated{0,1}
+  func iterAll<K>(t : Trie.Trie<K, ()>)
+    : Iter.Iter<K>
+    =
+    object {
+    var stack = ?(t, null) : List.List<Trie.Trie<K, ()>>;
+    public func next() : ?K {
+      switch stack {
+      case null { null };
+      case (?(trie, stack2)) {
+             switch trie {
+             case (#empty) {
```

Comment:
Nit: excessive indentation here, does this use tabs?

Reply:
No tabs.

Backstory: I'm using an emacs mode that insists of doing this to me.  It's the same emacs mode we have in the repo.  I've been hobbling along with it for two years and it's always been broken around switch/case indentation.